egld-esdt-swap/src/lib.rs

```diff
@@ -82,24 +105,75 @@ pub trait EgldEsdtSwap {
 
     // private
 
-    fn data_or_empty(&self, to: &Address, data: &'static [u8]) -> &[u8] {
-        if self.blockchain().is_smart_contract(to) {
-            &[]
-        } else {
-            data
-        }
+    fn needs_execution(&self, caller: &Address, function: &BoxedBytes) -> bool {
+        self.blockchain().is_smart_contract(caller) && !function.is_empty()
+    }
+
+    fn transfer_and_execute_via_async(
+        &self,
+        caller: &Address,
+        token_id: &TokenIdentifier,
+        amount: &Self::BigUint,
+        function: &BoxedBytes,
+    ) -> AsyncCall<Self::SendApi> {
+        let contract_call: ContractCall<Self::SendApi, ()> =
+            ContractCall::new(self.send(), caller.clone(), function.clone())
```

Comment:
Should keep some gas for the callback (use `with_gas_limit` method of `ContractCall`). This helps in the cases where the receiving function does some expensive computations (like writing storage). If the async call runs out of gas, the callback will not be executed, and the initial funds will still be lost.

Local mint/burn costs 250k gas, so reserving around 500k should be enough, but this needs some testing to be sure.

Reply:
Done